org-code-javabuilder/lib/src/main/java/org/code/javabuilder/CodeExecutionManager.java

```diff
@@ -0,0 +1,145 @@
+      OutputAdapter outputAdapter,
+      ExecutionType executionType,
+      List<String> compileList,
+      JavabuilderFileManager fileManager,
+      LifecycleNotifier lifecycleNotifier) {
+    this.fileLoader = fileLoader;
+    this.inputHandler = inputHandler;
+    this.outputAdapter = outputAdapter;
+    this.executionType = executionType;
+    this.compileList = compileList;
+    this.fileManager = fileManager;
+    this.lifecycleNotifier = lifecycleNotifier;
+    this.executionInProgress = false;
+  }
+
+  /** Executes code synchronously, ensuring that pre- and post-execution steps are performed. */
+  public void execute() {
+    try {
+      this.onPreExecute();
+      try {
+        // Run code builder
+        this.executionInProgress = true;
+        final CodeBuilderRunnable runnable =
+            new CodeBuilderRunnable(
+                this.fileLoader,
+                this.outputAdapter,
+                this.tempFolder,
+                this.executionType,
+                this.compileList);
+        runnable.run();
+      } catch (Throwable e) {
+        // Catch any throwable here to ensure that onPostExecute() is always called
+        LoggerUtils.logException(e);
+      }
+      this.onPostExecute();
+    } catch (InternalServerError e) {
+      LoggerUtils.logError(e);
+    }
+  }
+
+  /**
+   * Request the code execution to cleanup early (e.g. in the case of an impending timeout). Code
+   * execution may still run, but this allows for post execution steps to still occur before the
+   * environment is shut down.
+   */
+  public void requestEarlyExit() {
+    if (!this.executionInProgress) {
+      Logger.getLogger(MAIN_LOGGER).warning("Early exit requested before execution started");
+      return;
+    }
+    try {
+      this.onPostExecute();
+      this.executionInProgress = false;
+    } catch (InternalServerError e) {
+      LoggerUtils.logError(e);
+    }
+  }
+
+  /**
+   * Pre-execution steps: 1) create temporary folder, 2) Replace System.in/out with custom in/out
+   */
+  private void onPreExecute() throws InternalServerError {
+    // Create temp folder
+    try {
+      this.tempFolder = Files.createTempDirectory("tmpdir").toFile();
+    } catch (IOException e) {
+      throw new InternalServerError(InternalErrorKey.INTERNAL_EXCEPTION, e);
+    }
+
+    // Save System in/out and replace with custom in/out
+    this.systemInputStream = System.in;
+    this.systemOutputStream = System.out;
+    this.overrideInputStream = new InputRedirectionStream(this.inputHandler);
+    this.overrideOutputStream = new OutputPrintStream(this.outputAdapter);
+    System.setOut(this.overrideOutputStream);
+    System.setIn(this.overrideInputStream);
+  }
+
+  /**
+   * Post-execution steps: 1) Notify listeners, 2) clean up global resources, 3) clear temporary
+   * folder, 4) close custom in/out streams, 5) Replace System.in/out with original in/out
+   */
+  private void onPostExecute() throws InternalServerError {
```

Comment:
will there be any issues if onPostExecute() is called twice? Not sure if it's possible since we only call `requestEarlyExit` on a timeout, but want to double check.

Reply:
So ideally `this.executionInProgress` should guard against that, but I'm realizing I never set it to false if execution completes as normal, so I'll fix that

Reply:
Added a test case to verify this as well.